### ISSUE TITLE
Psalm exits when attempting to load an invalid file via reflection

### ISSUE THREAD

**Issue by leighman** [vimeo/psalm#3994]
Not sure if this is avoidable.

I make a method `final` in a parent class.
When psalm attempts to load a child class that overrides the method, via reflection (?), it causes a fatal PHP error.
```
Using reflection to locate file for Child
PHP Fatal error:  Cannot override final method Parent::delete() in /vagrant/application/xxx/child_yy.php on line 10
Fatal error: Cannot override final method Parent::delete() in /vagrant/application/xxx/child_yy.php on line 10
```

I would expect `MethodSignatureMismatch` which I'm pretty sure I've seen working correctly (in the non-reflection case?) for other methods

**Comment by weirdan** [comment]
Yeah, 'loading via reflection' means 'executing the code and inspecting it with PHP Reflection API'. Any fatal errors would result in a crash. 

One way to solve it would be to introduce a kind of user-supplied `SymbolLocator` that could take a symbol (class, function, const, etc) name and return the file path where it's defined.

**Comment by muglug** [comment]
> One way to solve it would be to introduce a kind of user-supplied SymbolLocator that could take a symbol (class, function, const, etc) name and return the file path where it's defined.

That would mean there'd be three different ways to introduce userland classes to Psalm:

- via your composer autoloader (default, doesn't required PHP Reflection)
- via a custom-provided autoloader in the config (requires PHP Reflection for all loaded classes)
- via a special `SymbolLocator` plugin (which would not, in and of itself, require PHP Reflection)

That last option feels a _little_ extraneous, given the issues it would help uncover are few and far between.

**Comment by weirdan** [comment]
It would be a little extraneous - for classlikes. For other kinds of symbols (functions, constants) it may still  simplify things (read: better support for shitty legacy code from 2000s). I had limited success by adding files containing free-standing functions and non-autoloaded classes as stubs, but that feels like a wrong thing to do.

**Comment by Ocramius** [comment]
Just stumbled into this myself, as I'm analysing a legacy codebase that uses CodeIgniter paradigms.

It is very much impossible to inspect any code through an autoloader, since the codebase is sprinkled with `die()`, `exit()` and `throw` statements that are in the source files (reads: codebase is not PSR-1 compliant).

If I understand it correctly, what we need here is something alike BetterReflection to look for symbols inside a directory structure? Is that what `SymbolLocator` would do?  Also, how would one plug it into psalm?

**Comment by weirdan** [comment]
> If I understand it correctly, what we need here is something alike BetterReflection to look for symbols inside a directory structure?

Does Codeigniter use non-autoloadable symbols (functions, non-class constants)? If it's only classes put into unpredictable places you may work this around by building classmap autoloader ([generator is built into Composer](https://github.com/composer/composer/blob/bd2f321527c963475068e47856547e15fb5992e7/src/Composer/Autoload/ClassMapGenerator.php#L272-L277)) and **not** registering custom autoloader in `psalm.xml`.

**Comment by Ocramius** [comment]
> Does Codeigniter use non-autoloadable symbols (functions, non-class constants)? 

Yes, most likely need a classmap approach, but also many files start with something like:

```php
if (! SOME_EXPRESSION) {
    exit();
}
```
